cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -284,12 +287,37 @@ private boolean runReadyJobs() throws IOException, DatasetManagementException {
     }
 
     private boolean isTriggerSatisfied(Job job) {
-      // TODO: implement trigger checking
-      return true;
+      Trigger trigger = job.getSchedule().getTrigger();
+      if (trigger instanceof TimeTrigger) {
+        // TimeTrigger is satisfied as soon as the Notification arrive, due to how the Notification is initially created
+        return true;
+      }
+      if (trigger instanceof PartitionTrigger) {
+        PartitionTrigger partitionTrigger = (PartitionTrigger) trigger;
+        int numPartitions = 0;
+        for (Notification notification : job.getNotifications()) {
+          String numPartitionsString = notification.getProperties().get("numPartitions");
+          numPartitions += Integer.valueOf(numPartitionsString);
+        }
+        return numPartitions >= partitionTrigger.getNumPartitions();
+      }
+      throw new IllegalArgumentException("Unknown trigger class: " + trigger.getClass());
     }
 
     private boolean constraintsSatisfied(Job job) {
-      // TODO: implement constraint checking
+      for (Constraint constraint : job.getSchedule().getConstraints()) {
+        if (!(constraint instanceof AbstractCheckableConstraint)) {
+          // this shouldn't happen, since all Constraint implementations should extend AbstractConstraint
+          throw new IllegalArgumentException("Implementation of Constraint must extend AbstractConstraint");
+        }
+        AbstractCheckableConstraint abstractConstraint = (AbstractCheckableConstraint) constraint;
+        ConstraintResult result = abstractConstraint.check(job.getSchedule(), new ConstraintContext(job));
+        if (result != ConstraintResult.SATISFIED) {
```

Comment:
I thought in the very beginning we agreed that the `ProgramSchedule` would only carry one `Constraint` field, which it will be the constraint to check or a `AndConstraint` which "and" together multiple constraints. This allow future enhancement much easier.